pkg/store/consul/rcstore/consul_store.go

```diff
@@ -195,44 +197,78 @@ func (s *consulStore) WatchNew(quit <-chan struct{}) (<-chan []fields.RC, <-chan
 	return outCh, errCh
 }
 
-// Wraps an RC with lock information
+// Wraps an RC ID with lock information
+// This structure is useful for low-bandwidth detection of new RCs, for use
+// in the RC farm.
 type RCLockResult struct {
-	fields.RC
+	ID                      fields.ID
 	LockedForOwnership      bool
 	LockedForUpdateCreation bool
 	LockedForMutation       bool
 }
 
-// Like WatchNew() but instead of returning raw fields.RC types it wraps them
-// in a struct that indicates what types of locks are held on the RC. This is
-// useful in reducing failed lock attempts in the RC farms which have a latency
-// cost. WatchNewWithLockInfo() can retrieve the contents of the entire
-// replication controller lock tree once and farms can only attempt to acquire
-// locks that were not held at some recent time
+// WatchRCKeysWithLockInfo executes periodic consul watch queries for
+// replication controller keys and passes the results on a channel along with
+// information about which keys have locks on them. Only keys are returned as a
+// bandwidth performance optimization. To observe changes to the fields of a
+// replication controller, individual watches should be used. Lock info is
+// returned so that it can be used to not try to lock RCs that are already
+// locked. In other words, it's a QPS optimization over the naive solution of
+// always attempting to lock every RC and relying on the consul server to
+// decide whether the lock is own.
 //
 // Since lock information is retrieved once per update to the RC list, it's
 // possible that lock information will be out of date as the list is processed.
 // However, a subsequent update will get the correct view of the world so the
 // behavior should be correct
-func (s *consulStore) WatchNewWithRCLockInfo(quit <-chan struct{}, pauseTime time.Duration) (<-chan []RCLockResult, <-chan error) {
-	inCh := make(chan api.KVPairs)
-	lockInfoErrCh := make(chan error)
+func (s *consulStore) WatchRCKeysWithLockInfo(quit <-chan struct{}, pauseTime time.Duration) (<-chan []RCLockResult, <-chan error) {
 	combinedErrCh := make(chan error)
 
-	rcCh, rcErrCh := publishLatestRCs(inCh, quit)
-	go consulutil.WatchPrefix(rcTree+"/", s.kv, inCh, quit, rcErrCh, pauseTime)
+	keyCh := consulutil.WatchKeys(rcTree+"/", s.kv, quit, pauseTime)
+
+	var combinedErrChWG sync.WaitGroup
+	rcIDCh := make(chan []fields.ID)
+	combinedErrChWG.Add(1)
+	go func() {
+		defer combinedErrChWG.Done()
+		for watchedKeys := range keyCh {
+			if watchedKeys.Err != nil {
+				select {
+				case <-quit:
+					return
+				case combinedErrCh <- watchedKeys.Err:
+					continue
+				}
+			}
+
+			rcIDs, err := keysToRCIDs(watchedKeys.Keys)
+			if err != nil {
+				select {
+				case <-quit:
+					return
+				case combinedErrCh <- watchedKeys.Err:
```

Comment:
(I predict this comment will end up displaying on the wrong line - I mean to comment on line 249, after you check `err != nil`)

Checking `err != nil`, but sending `watchedKeys.Err` instead of `err`? Should it be `err` instead? If it's intentional, tell me why